generator/restriction_collector.cpp

```diff
@@ -0,0 +1,210 @@
+#include "generator/restriction_collector.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+#include "base/stl_helpers.hpp"
+#include "base/string_utils.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/fstream.hpp"
+
+namespace
+{
+char const kNo[] = "No";
+char const kOnly[] = "Only";
+
+bool ParseLineOfNumbers(strings::SimpleTokenizer & iter, vector<uint64_t> & numbers)
+{
+  uint64_t number = 0;
+  while (iter)
+  {
+    if (!strings::to_uint64(*iter, number))
+      return false;
+    numbers.push_back(number);
+    ++iter;
+  }
+  return true;
+}
+}  // namespace
+
+namespace routing
+{
+RestrictionCollector::RestrictionCollector(string const & restrictionPath,
+                                           string const & featureId2OsmIdsPath)
+{
+  if (!ParseFeatureId2OsmIdsMapping(featureId2OsmIdsPath))
+    return;
+
+  if (!ParseRestrictions(restrictionPath))
+  {
+    m_restrictions.clear();
+    return;
+  }
+  ComposeRestrictions();
+  RemoveInvalidRestrictions();
+  LOG(LINFO, ("m_restrictions.size() =", m_restrictions.size()));
+}
+
+bool RestrictionCollector::IsValid() const
+{
+  return !m_restrictions.empty() &&
+         find_if(begin(m_restrictions), end(m_restrictions),
+                 [](Restriction const & r) { return !r.IsValid(); }) == end(m_restrictions);
+}
+
+bool RestrictionCollector::ParseFeatureId2OsmIdsMapping(string const & featureId2OsmIdsPath)
+{
+  ifstream featureId2OsmIdsStream(featureId2OsmIdsPath);
+  if (featureId2OsmIdsStream.fail())
+    return false;
+
+  while (featureId2OsmIdsStream)
+  {
+    string line;
+    if (!getline(featureId2OsmIdsStream, line))
+      return true;
+
+    vector<uint64_t> osmIds;
+    strings::SimpleTokenizer iter(line, ",");
+    if (!ParseLineOfNumbers(iter, osmIds))
+      return false;
+
+    if (osmIds.size() < 2)
+      return false;  // Every line should contain at least feature id and osm id.
+
+    uint32_t const featureId = static_cast<uint32_t>(osmIds.front());
+    osmIds.erase(osmIds.begin());
+    AddFeatureId(featureId, osmIds);
+  }
+  return true;
+}
+
+bool RestrictionCollector::ParseRestrictions(string const & restrictionPath)
+{
+  ifstream restrictionsStream(restrictionPath);
+  if (restrictionsStream.fail())
+    return false;
+
+  while (restrictionsStream)
+  {
+    string line;
+    if (!getline(restrictionsStream, line))
+      return true;
+
+    strings::SimpleTokenizer iter(line, ",");
+    Restriction::Type type;
+    if (!FromString(*iter, type))
```

Comment:
Did you mean to add this check before "if (!FromString(*iter, type))"? If so, done.